internal/cluster/cluster.go

```diff
@@ -522,32 +524,52 @@ func (m *Manager) ClusterMonitoring() {
 		if len(clusters) == 0 {
 			break
 		}
-		for _, cluster := range clusters {
-			if !m.leaderElector.IsLeader() {
-				m.log.Debugf("Not a leader, exiting ClusterMonitoring")
-				return
-			}
-			if !m.SkipMonitoring(cluster) {
 
-				if err = m.setConnectivityMajorityGroupsForClusterInternal(cluster, m.db); err != nil {
-					log.WithError(err).Error("failed to set majority group for clusters")
-				}
-				clusterAfterRefresh, err = m.refreshStatusInternal(ctx, cluster, m.db)
-				if err != nil {
-					log.WithError(err).Errorf("failed to refresh cluster %s state", cluster.ID)
-					continue
-				}
-
-				if swag.StringValue(clusterAfterRefresh.Status) != swag.StringValue(cluster.Status) {
-					log.Infof("cluster %s updated status from %s to %s via monitor", cluster.ID,
-						swag.StringValue(cluster.Status), swag.StringValue(clusterAfterRefresh.Status))
-				}
+		jobs := make(chan *common.Cluster, limit)
+		for i := range clusters {
+			jobs <- clusters[i]
+		}
 
-				if m.shouldTriggerLeaseTimeoutEvent(cluster, curMonitorInvokedAt) {
-					m.triggerLeaseTimeoutEvent(ctx, cluster)
+		var wg sync.WaitGroup
+		for i := 0; i < m.MonitorWorkersNum; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for {
+					select {
+					case cluster := <-jobs:
+						if !m.leaderElector.IsLeader() {
+							m.log.Debugf("Not a leader, exiting ClusterMonitoring")
+							return
+						}
+						if !m.SkipMonitoring(cluster) {
+
+							if err = m.setConnectivityMajorityGroupsForClusterInternal(cluster, m.db); err != nil {
+								log.WithError(err).Error("failed to set majority group for clusters")
+							}
+							clusterAfterRefresh, err = m.refreshStatusInternal(ctx, cluster, m.db)
+							if err != nil {
+								log.WithError(err).Errorf("failed to refresh cluster %s state", cluster.ID)
+								continue
+							}
+
+							if swag.StringValue(clusterAfterRefresh.Status) != swag.StringValue(cluster.Status) {
+								log.Infof("cluster %s updated status from %s to %s via monitor", cluster.ID,
+									swag.StringValue(cluster.Status), swag.StringValue(clusterAfterRefresh.Status))
+							}
+
+							if m.shouldTriggerLeaseTimeoutEvent(cluster, curMonitorInvokedAt) {
+								m.triggerLeaseTimeoutEvent(ctx, cluster)
+							}
+						}
+					default:
```

Comment:
I think default is wrong in this case. Maybe the channel is empty temporarily

Reply:
the channel is set before the workers are activated.

Reply:
But if the channel is empty temporarily, the default clause will be called.  Doesn't seem what you want.  Isn't it?

Reply:
This way workers will terminate before the batch is completed.

Reply:
I don't understand how. 
Workers will start only after the channel is filled with jobs.

Reply:
Changed the flow, now workers are running during all iterations and not create for each batch.